libsplinter/src/admin/store/diesel/operations/count_circuits.rs

```diff
@@ -0,0 +1,118 @@
+    i32: diesel::deserialize::FromSql<diesel::sql_types::Integer, C::Backend>,
+    i16: diesel::deserialize::FromSql<diesel::sql_types::SmallInt, C::Backend>,
+{
+    fn count_circuits(
+        &self,
+        predicates: &[CircuitPredicate],
+    ) -> Result<u32, AdminServiceStoreError> {
+        // Collect the management types included in the list of `CircuitPredicates`
+        let management_types: Vec<String> = predicates
+            .iter()
+            .filter_map(|pred| match pred {
+                CircuitPredicate::ManagementTypeEq(man_type) => Some(man_type.to_string()),
+                _ => None,
+            })
+            .collect::<Vec<String>>();
+        // Collects the members included in the list of `CircuitPredicates`
+        let members: Vec<String> = predicates
+            .iter()
+            .filter_map(|pred| match pred {
+                CircuitPredicate::MembersInclude(members) => Some(members.to_vec()),
+                _ => None,
+            })
+            .flatten()
+            .collect();
+        let statuses: Vec<CircuitStatusModel> = predicates
+            .iter()
+            .filter_map(|pred| match pred {
+                CircuitPredicate::CircuitStatus(status) => Some(CircuitStatusModel::from(status)),
+                _ => None,
+            })
+            .collect();
+        self.conn.transaction::<u32, _, _>(|| {
+            // Collects circuits which match the circuit predicates
+            let mut query = circuit::table.into_boxed().select(circuit::all_columns);
+
+            if !management_types.is_empty() {
+                query = query.filter(circuit::circuit_management_type.eq_any(management_types));
+            }
+
+            if !members.is_empty() {
+                query = query.filter(exists(
+                    // Selects all `circuit_member` entries where the `node_id` is equal
+                    // to any of the members in the circuit predicates
+                    circuit_member::table.filter(
+                        circuit_member::circuit_id
+                            .eq(circuit::circuit_id)
+                            .and(circuit_member::node_id.eq_any(members)),
+                    ),
+                ));
+            }
+
+            if statuses.is_empty() {
+                // By default, only display active circuits
+                query = query.filter(circuit::circuit_status.eq(CircuitStatusModel::Active));
+            } else {
+                query = query.filter(
+                    // Select only circuits that have the `CircuitStatus` in the predicates
+                    circuit::circuit_status.eq_any(statuses),
+                );
+            }
+
+            let count = query.select(count_star()).first::<i64>(self.conn)?;
+
+            u32::try_from(count).map_err(|_| {
+                AdminServiceStoreError::InternalError(InternalError::with_message(
+                    "The number of proposals is larger than the max u32".to_string(),
```

Comment:
Looks like some copy-pasta in this error message

Reply:
fixed